### ISSUE TITLE
POST request with non JSON content type data

### ISSUE THREAD

**Issue by ribeaud** [svanoort/pyresttest#115]
It seems that the interface I am interacting with does NOT support content type **application/json** for `POST` requests. After a brief analysis, I got the impression that it is NOT possible to submit *data* to `POST` requests in a format different than **application/json**. Am I right?
You submit data via:
```
curl.setopt(curl.READFUNCTION, StringIO(bod).read)
```
and NOT via:
```
 curl.setopt(pycurl.POSTFIELDS, urlencode(body))
```
although it is possible to specify a `dict` as *body* structure. It would be nice if we have more flexibility here. I will pull a request after having made the corresponding changes.

**Comment by svanoort** [comment]
@ribeaud 
This may be the first time you're happy to find you're wrong!   PyRestTest supports submitting everything from JSON to images, etc, however admittedly doing form data this way could be made a lot easier (I'm converting this issue into a feature story to address that).

The reason for this is that PyRestTest takes a very-level approach to requests: from its point of view content type is simply a header set along with whatever supplied binary content is used for the request body.  I am currently adding some encoding awareness to the content types so that if you supply Unicode text without applying an encoding, it will set the Unicode content type and encode it.

So, for a PyRestTest body, this would be how you submit the following form data:

```python
data = {'Name': 'Gareth Wylie',
  'Age': 24,
  'Formula': 'a + b == 13%!'
}
```

**Test syntax:**
```yaml
- test:
  - name: 'POST some form data!'
  - method: 'POST'
  - body: 'Name=Gareth+Wylie&Age=24&Formula=a+%2B+b+%3D%3D+13%25%21'
  - headers: {'Content-Type': 'application/x-www-form-urlencoded'}
``

[Standard reference on this use of POST for form data those less familiar than you or I who may see it.](https://en.wikipedia.org/wiki/POST_(HTTP)#Use_for_submitting_web_forms)

Worth noting also that you also have the escape value of [Custom Curl Options](https://github.com/svanoort/pyresttest#custom-http-options-special-curl-settings)  to override the way we use any of the Curl settings.  This sometimes will offer a quick way to accomplish very custom or unusual scenarios.

**Comment by svanoort** [comment]
@ribeaud So, here's how I'd propose to convert this to a helpful feature in the next release or two, tell me what you think:

**Implementation:**

Add a 'form' element to the ContentHandler class, so that you can supply, read from file, and template out form data as a request body.  This is internally URL encoded and concatenated with separators per the HTTP spec, as well as setting the default header (unless you choose to override this).


**Test syntax:**

```yaml
- test:
  - name: "I submit form data'
  - method: POST
  - body: {form: {'parameter': 'value', 'cheese': 'gouda', 'foo': 'bar'}}
```

And it would do all the above implicitly under the hood in the configure_curl method (or a helper method more likely, to keep that short).

With this you'd get template/variable binding integration, the ability to read from file, current encoding handling goodies, etc.  The catch is that it adds one more dimension to the combinatorial matrix that is the ContentHandler.

**Timeline:**

Currently I'm in the midst of major revisions to support Python 3, handle encoding more safely, and offer better support for dynamic content passing.  These would all drive potentially significant changes to the same code, so this is going to come after that (unless someone in the community has time to put together a PR).  

Since there's a way to do it currently (albeit clumsily) and side project time is limited at the moment, probably we're looking at the release *after* next unless I get a burst of time or a PR comes in.

**Comment by ribeaud** [comment]
Hi @svanoort. Works like a charm. And yes, I am quite happy, I do not have to tag the library... :)

I like your proposal. I had equivalent idea. If I would have time, I would even go a bit further, offering a *registering* infrastructure the same way you did for **validators** with `register_validator`. We would have one implementation for *file*, *form* resp. *template* namespace. And a default one which throws an exception:
```
raise Exception("Invalid configuration for content.")
```
So people would be able to extend this in a plugin-like way.

Until it is implemented, I would add an example in the documentation just like you did for me, demonstrating that `application/x-www-form-urlencoded` posting is possible as well. Maybe I did not read the documentation carefully enough but I got the impression that only **json** submitting was possible. Otherwise I would not have disturbed you...

**Comment by svanoort** [comment]
Yes, and yes to both!  (Thought adding that as an extension point is more complex than one would anticipate, unfortunately.)

**Comment by svanoort** [comment]
@ribeaud Example & link in README added with commit https://github.com/svanoort/pyresttest/commit/fa57dd0dc8e12469d5e93b3af0665f725ac2e3b0
